### PR TITLE
FailingPartitions throw exception if fetched StreamEvent is not in expected partition

### PR DESCRIPTION
**Pull request by joelhoisko**
Should be a rare case as something really has to mess up for this to happen but better safe than sorry.

Multilined a lot of long lines too.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1140654762299587/1174933939820338)

### CODE REVIEW

File: Source/Events.Processing/Streams/Partitioned/ExpectedStreamEventWithSameGivenPartition.cs [hunk 1]
```diff
@@ -0,0 +1,24 @@
+// Copyright (c) Dolittle. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using Dolittle.Runtime.Events.Store.Streams;
+
+namespace Dolittle.Runtime.Events.Processing.Streams.Partitioned
+{
+    /// <summary>
+    /// Exception that gets thrown when the <see cref="StreamEvent"/> has a different <see cref="PartitionId"/> as expected.
```
Comment (woksin) [inline Source/Events.Processing/Streams/Partitioned/ExpectedStreamEventWithSameGivenPartition.cs:10]:
```suggestion
    /// Exception that gets thrown when a <see cref="StreamEvent"/> has a different <see cref="PartitionId"/> than expected.
```

File: Source/Events.Processing/Streams/Partitioned/ExpectedStreamEventWithSameGivenPartition.cs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    /// </summary>
+    public class ExpectedStreamEventWithSameGivenPartition : Exception
```
Comment (woksin) [inline Source/Events.Processing/Streams/Partitioned/ExpectedStreamEventWithSameGivenPartition.cs:12]:
```suggestion
    public class StreamEventInWrongPartition : Exception
```

File: Source/Events.Processing/Streams/Partitioned/ExpectedStreamEventWithSameGivenPartition.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ExpectedStreamEventWithSameGivenPartition"/> class.
```
Comment (woksin) [inline Source/Events.Processing/Streams/Partitioned/ExpectedStreamEventWithSameGivenPartition.cs:15]:
```suggestion
        /// Initializes a new instance of the <see cref="StreamEventInWrongPartition"/> class.
```

File: Source/Events.Processing/Streams/Partitioned/ExpectedStreamEventWithSameGivenPartition.cs [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+        /// </summary>
+        /// <param name="event">The <see cref="StreamEvent"/> with the wrong <see cref="PartitionId"/>.</param>
+        /// <param name="givenPartition">The expected <see cref="PartitionId"/>.</param>
+        public ExpectedStreamEventWithSameGivenPartition(StreamEvent @event, PartitionId givenPartition)
```
Comment (woksin) [inline Source/Events.Processing/Streams/Partitioned/ExpectedStreamEventWithSameGivenPartition.cs:19]:
```suggestion
        public StreamEventInWrongPartition(StreamEvent @event, PartitionId expectedPartitionId)
```

File: Source/Events.Processing/Streams/Partitioned/ExpectedStreamEventWithSameGivenPartition.cs [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+            : base($"Expected StreamEvent {@event} to have the same PartitionId as {givenPartition}")
```
Comment (woksin) [inline Source/Events.Processing/Streams/Partitioned/ExpectedStreamEventWithSameGivenPartition.cs:20]:
```suggestion
            : base($"Expected Event at Position: {@event.Position} in Stream: '{@event.Stream}' to be in Partition: '{expectedPartitionId}' but was in Partition: {@event.Partition}")
```

File: Source/Events.Processing/Streams/Partitioned/FailingPartitions.cs [hunk 6]
```diff
@@ -66,26 +81,56 @@ public async Task<IStreamProcessorState> CatchupFor(IStreamProcessorId streamPro
                 if (ShouldRetryProcessing(failingPartitionState))
                 {
                     var streamEvent = await _eventsFromStreamsFetcher.FetchInPartition(partition, failingPartitionState.Position, cancellationToken).ConfigureAwait(false);
+                    if (streamEvent.Partition != partition) throw new ExpectedStreamEventWithSameGivenPartition(streamEvent, partition);
```
Comment (woksin) [inline Source/Events.Processing/Streams/Partitioned/FailingPartitions.cs:84]:
```suggestion
                    if (streamEvent.Partition != partition) throw new StreamEventInWrongPartition(streamEvent, partition);
```

File: Source/Events.Processing/Streams/Partitioned/ExpectedStreamEventWithSameGivenPartition.cs (hunk 1 above)
Comment (joelhoisko) [reply Source/Events.Processing/Streams/Partitioned/ExpectedStreamEventWithSameGivenPartition.cs]: Fixed

File: Source/Events.Processing/Streams/Partitioned/ExpectedStreamEventWithSameGivenPartition.cs (hunk 3 above)
Comment (joelhoisko) [reply Source/Events.Processing/Streams/Partitioned/ExpectedStreamEventWithSameGivenPartition.cs]: Renamed the class

**Pull request merged**